Unity-Technologies/ScriptableRenderPipeline
# Various blackboard improvements

Finishing up a "20%" branch; various improvements for the blackboard;

* Blackboard Properties now have a Duplication popup menu option (https://favro.com/organization/c564ede4ed3337f7b17986b6/d06b8de3e40da6d340b5b063?card=Uni-74375)
![DuplicateBlackboardOption0](https://user-images.githubusercontent.com/3393041/69021656-67749400-096d-11ea-834a-6040a3f65ba7.gif)
* Bug fix: exposed is now properly duplicated with ShaderInputs
* Bug fix: duplicating multiple properties at once often crashed the entire graph https://fogbugz.unity3d.com/f/cases/1195157/ (see BlackboardProvider.cs)
* Blackboard Properties are now pasted in order (as opposed to randomly):
![InOrder0](https://user-images.githubusercontent.com/3393041/69024184-84fa2b80-0976-11ea-925d-9ab0cc406383.gif)
* Duplicated Properties are now pasted below the last selected blackboard property, as opposed to the end of the section (also displayed by the above gif)
* [Nevermind this text: it has already been fixed outside of this PR] Bug fix: https://fogbugz.unity3d.com/f/cases/1189065/ **[!]** Need discussion, as mentioned in the bug case it was introduce by https://github.com/Unity-Technologies/ScriptableRenderPipeline/commit/eca116d0f313223ab9c6a4dd77d10addb4d999b3. Unsure of reason for sorting by guid, seems just to be for human readability for otherwise everything functions the same without. Need fix for above improvement.
* Fixed an unreported bug where Built-In Keywords could be duplicated via Blackboard Duplication:
![BuiltinKeywordsNotDuplicated0](https://user-images.githubusercontent.com/3393041/69024433-909a2200-0977-11ea-8e67-3db6af9a18e3.gif)

---

Halo Effect: 3/4
Technical Risk: 1/4

---
### Testing status

**Manual Tests**: What did you do?
- Duplicated many different ShaderInputs, including within the same graph, between graphs, and sub graphs.
- Messing with order of selection before duplication
- Existing graphs, new graphs
- Various Shader Inputs, duplication both Properties & Keywords simultaneously
- All combinations with means of duplication (ctrl+d, copy & paste, Duplication menu option)

---

_Misc:_

* Removed controversial change that made exposed serialized, see #5046
* Will talk to fellow qa about 2nd qa pass
* Commits allow certain features of this PR to be omitted

---

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline/tree/sg%252Ffix%252F1187828/.yamato%252Fupm-ci-shadergraph.yml%2523All_ShaderGraph_trunk
(Master compare: https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline/tree/master/.yamato%252Fupm-ci-packages.yml%2523all_package_ci/1009336/job)
